### PR TITLE
Fixed the version bug and also added force flag

### PR DESCRIPTION
**Pull request by pmahindrakar-oss**
Signed-off-by: Prafulla Mahindrakar <prafulla.mahindrakar@gmail.com>

# TL;DR
Fixes the version substitution bug

If serialization has generated version identifier to be used for various flyte entities then those should be used instead of passed in version flag.

An additional force flag has been introduced which overrides the above behavior and uses the passed in version for all registerable flyte entities irrespective of there serialization status

Testing done  : 

####  Passed version flag but no force flag . Entities would be registered with serialized version
```
flytectl register file  /Users/praful/Downloads/out.tar.gz  -d development  -p chariot-sdk-test  --version v1 --logger.level 2 --archive 
 ---------------------------------------------------------------------- --------- ------------------------------ 
| NAME (4)                                                             | STATUS  | ADDITIONAL INFO              |
 ---------------------------------------------------------------------- --------- ------------------------------ 
| /tmp/register407864907/0_pm.nb.new.ipynb_1.pb                        | Success | Successfully registered file |
 ---------------------------------------------------------------------- --------- ------------------------------ 
| /tmp/register407864907/1_new.ipynb_1.pb                              | Success | Successfully registered file |
 ---------------------------------------------------------------------- --------- ------------------------------ 
| /tmp/register407864907/2_workflows.new_workflow.nb_to_python_wf_2.pb | Success | Successfully registered file |
 ---------------------------------------------------------------------- --------- ------------------------------ 
| /tmp/register407864907/3_workflows.new_workflow.nb_to_python_wf_3.pb | Success | Successfully registered file |
 ---------------------------------------------------------------------- --------- ------------------------------ 
```

#### Flyte entities registered with serialized version 1

```
flytectl get task -p chariot-sdk-test -d development --logger.level 6

 --------- ----------------- ---------------- -------- ----------------- -------------- ------------------- ----------------------------- 
| VERSION | NAME            | TYPE           | INPUTS | OUTPUTS         | DISCOVERABLE | DISCOVERY VERSION | CREATED AT                  |
 --------- ----------------- ---------------- -------- ----------------- -------------- ------------------- ----------------------------- 
| 1       | new.ipynb       | nb-python-task | a      | message         |              |                   | 2021-12-22T12:00:05.633777Z |
|         |                 |                | b      | out_nb          |              |                   |                             |
|         |                 |                |        | out_rendered_nb |              |                   |                             |
|         |                 |                |        | thrice          |              |                   |                             |
|         |                 |                |        | twice           |              |                   |                             |
 --------- ----------------- ---------------- -------- ----------------- -------------- ------------------- ----------------------------- 
| 1       | pm.nb.new.ipynb | python-task    |        |                 |              |                   | 2021-12-22T12:00:05.626048Z |
 --------- ----------------- ---------------- -------- ----------------- -------------- ------------------- ----------------------------- 
```

#### Registration with new passed in version but fails (Confusing at first since version 1 already exists and no version v2 exists for any flyte entities and --version v2 is ignored)

```
flytectl register file  /Users/praful/Downloads/out.tar.gz  -d development  -p chariot-sdk-test  --version v2 --logger.level 2 --archive        
{"json":{"src":"client.go:170"},"level":"error","msg":"failed to initialize token source provider. Err: rpc error: code = Unimplemented desc = unknown service flyteidl.service.AuthMetadataService","ts":"2021-12-22T17:31:25+05:30"}
 ----------------------------------------------------------------------- --------- ----------------- 
| NAME (4)                                                              | STATUS  | ADDITIONAL INFO |
 ----------------------------------------------------------------------- --------- ----------------- 
| /tmp/register1866986520/0_pm.nb.new.ipynb_1.pb                        | Success | AlreadyExists   |
 ----------------------------------------------------------------------- --------- ----------------- 
| /tmp/register1866986520/1_new.ipynb_1.pb                              | Success | AlreadyExists   |
 ----------------------------------------------------------------------- --------- ----------------- 
| /tmp/register1866986520/2_workflows.new_workflow.nb_to_python_wf_2.pb | Success | AlreadyExists   |
 ----------------------------------------------------------------------- --------- ----------------- 
| /tmp/register1866986520/3_workflows.new_workflow.nb_to_python_wf_3.pb | Success | AlreadyExists   |
 ----------------------------------------------------------------------- --------- ----------------- 
```

#### Force registration with v1

```
flytectl register file  /Users/praful/Downloads/out.tar.gz  -d development  -p chariot-sdk-test  --version v1 --logger.level 2 --archive --force
 ----------------------------------------------------------------------- --------- ------------------------------ 
| NAME (4)                                                              | STATUS  | ADDITIONAL INFO              |
 ----------------------------------------------------------------------- --------- ------------------------------ 
| /tmp/register2385553692/0_pm.nb.new.ipynb_1.pb                        | Success | Successfully registered file |
 ----------------------------------------------------------------------- --------- ------------------------------ 
| /tmp/register2385553692/1_new.ipynb_1.pb                              | Success | Successfully registered file |
 ----------------------------------------------------------------------- --------- ------------------------------ 
| /tmp/register2385553692/2_workflows.new_workflow.nb_to_python_wf_2.pb | Success | Successfully registered file |
 ----------------------------------------------------------------------- --------- ------------------------------ 
| /tmp/register2385553692/3_workflows.new_workflow.nb_to_python_wf_3.pb | Success | Successfully registered file |
 ----------------------------------------------------------------------- --------- ------------------------------ 
```

#### Get workflows

```
flytectl get workflow -p chariot-sdk-test -d development --logger.level 6
 --------- ---------------------------------------- ----------------------------- 
| VERSION | NAME                                   | CREATED AT                  |
 --------- ---------------------------------------- ----------------------------- 
| v1      | workflows.new_workflow.nb_to_python_wf | 2021-12-22T12:01:05.244326Z |
 --------- ---------------------------------------- ----------------------------- 
| 1       | workflows.new_workflow.nb_to_python_wf | 2021-12-22T12:00:05.653426Z |
 --------- ---------------------------------------- ----------------------------- 
```

## Type
- [X] Bug Fix
- [X] Feature
- [ ] Plugin

## Are all requirements met?

- [X] Code completed
- [X] Smoke tested
- [ ] Unit tests added
- [ ] Code documentation added
- [ ] Any pending items have an associated Issue

## Complete description
_How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/1970

## Follow-up issue
_NA_

### CODE REVIEW

File: cmd/register/register_util.go [hunk 1]
```diff
@@ -315,29 +315,30 @@ func hydrateSpec(message proto.Message, sourceCode string, config rconfig.FilesC
 	switch v := message.(type) {
 	case *admin.LaunchPlan:
 		launchPlan := message.(*admin.LaunchPlan)
-		hydrateIdentifier(launchPlan.Spec.WorkflowId, config.Version)
```
Comment (evalsocket) [inline cmd/register/register_util.go:318]: I think earlier we decided that we will no use config directly, We will use parameter.

**Review by evalsocket — APPROVED** [review]
